### DISCUSSION TITLE
Connection handling for Sql4CdsConnection

### DISCUSSION THREAD

**Discussion by jarvan-jiang** [MarkMpn/Sql4Cds#568] [General]
Hi Mark, 

I noticed that the code doesn’t explicitly call Open or Close methods for Sql4CdsConnection. Does this mean the connection isn't technically "open" and therefore doesn’t require closing?

The reason I’m asking is that I've observed some queries take a few seconds initially when opening the connection. Using it like a regular SQL connection introduces latency due to frequent open/close operations.

Would it be possible to keep the connection open, perhaps as a singleton? Or do you have any recommended patterns to reduce this overhead?

**Comment by MarkMpn** [comment]
`Sql4CdsConnection` is a wrapper around the underlying `IOrganizationService` - you can keep a reference to the same connection for the lifetime of the `IOrganizationService`.

If you are managing multiple connections to the same instance you could also look at using shared `IAtrributeMetadataCache`, `ITableSizeCache` and `IMessageCache` instances and passing them to the `DataSource` constructor rather than letting SQL 4 CDS create new ones each time - the majority of the startup time will be taken by populating these caches.

**Reply by jarvan-jiang** [reply]
Thanks Mark! Manage `DataSource` by myself did the trick!
